Ngcc migrations

petebacondarwin opened the pull request:
Relevant design docs:

https://hackmd.io/KhyrFV1VQHmeQsgfJq6AyQ?both
https://hackmd.io/sfb3Ju2MTmKHSUiX_dLWGg

TODO:

- [ ] how to deal with migration errors
- [ ] relax `@Directive()` so that it doesn't need a `selector`

alxhub reviewed (changes requested):
A few minor comments, but this looking super good!

petebacondarwin commented on packages/compiler-cli/ngcc/src/migrations/migration.ts:
```diff
@@ -0,0 +1,23 @@
+
+export interface MigrationHost {
+  metadata: MetadataReader;
+  host: NgccReflectionHost;
```
Yeah I was starting to think that myself :-)
`host.host.host....`

petebacondarwin commented on packages/compiler-cli/ngcc/src/analysis/migration_host.ts:
```diff
@@ -0,0 +1,57 @@
+    if (oldAnalyzedClass === undefined) {
+      analyzedFile.analyzedClasses.push(newAnalyzedClass);
+    } else {
+      // TBD: merge the analysis!!
```
To clarify - we will error if this call attempts to inject a decorator that already exists on the class.

petebacondarwin commented on packages/compiler-cli/ngcc/src/migrations/undecorated_parent_migration.ts:
```diff
@@ -0,0 +1,85 @@
+      return null;
+    }
+
+    const baseClazz = host.host.getDeclarationOfIdentifier(baseClassId) !.node as ClassDeclaration;
```
OK but it is a bit ugly:

```typescript
    if (!isNamedClassDeclaration(baseClazz) && !isNamedFunctionDeclaration(baseClazz) &&
        !isNamedVariableDeclaration(baseClazz)) {
      return null;
    }
```

gkalpak commented on packages/compiler-cli/ngcc/BUILD.bazel:
```diff
@@ -14,6 +14,7 @@ ts_library(
         "//packages/compiler",
         "//packages/compiler-cli/src/ngtsc/annotations",
         "//packages/compiler-cli/src/ngtsc/cycles",
+        "//packages/compiler-cli/src/ngtsc/diagnostics",
```
Doesn't this change belong to the previous commit, which adds support for migrations?

petebacondarwin replied:
Yes! Absolutely.

gkalpak commented on packages/compiler-cli/ngcc/src/migrations/undecorated_parent_migration.ts:
```diff
@@ -0,0 +1,128 @@
+      return null;
+    }
+
+    // Only interested in this base class if it doesn't have a `Directive` or Component` decorator.
```
Typo: `` Component` `` --> `` `Component` ``

gkalpak commented on packages/compiler-cli/ngcc/src/migrations/undecorated_parent_migration.ts:
```diff
@@ -0,0 +1,128 @@
+    const baseClazz = host.reflectionHost.getDeclarationOfIdentifier(baseClassId) !.node;
+
+    if (!isNamedClassDeclaration(baseClazz) && !isNamedFunctionDeclaration(baseClazz) &&
+        !isNamedVariableDeclaration(baseClazz)) {
```
It might be worth having a helper combining these (if it's going to be useful in other migrations).

petebacondarwin replied:
In classic XP style, we can add that when someone (👀) implements the next migration :-)

petebacondarwin replied:
Along with moving some of the helpers below into a common util file, maybe?

gkalpak reviewed (approved):
A few minor comments. Looks great otherwise 👍

petebacondarwin commented on packages/compiler-cli/ngcc/src/analysis/migration_host.ts:
```diff
@@ -0,0 +1,81 @@
+
+function getOrCreateAnalyzedFile(
+    analyzedFiles: AnalyzedFile[], sourceFile: ts.SourceFile): AnalyzedFile {
+  const analyzedFileIndex = analyzedFiles.findIndex(file => file.sourceFile === sourceFile);
```
I think a previous version of the code did more. But I think you are right now.

petebacondarwin commented on packages/compiler-cli/ngcc/test/host/esm2015_host_spec.ts:
```diff
@@ -1680,6 +1680,52 @@ runInEachFileSystem(() => {
 
+    describe('getBaseClassIdentifier()', () => {
+      it('should not consider a class without extends clause as having a base class', () => {
+        const file = {
```
This is following the same pattern as `hasBaseClass()` tests above.

petebacondarwin commented on packages/compiler-cli/ngcc/test/host/esm2015_host_spec.ts:
```diff
@@ -1680,6 +1680,52 @@ runInEachFileSystem(() => {
+        const {program} = makeTestBundleProgram(file.name);
+        const host = new Esm2015ReflectionHost(new MockLogger(), false, program.getTypeChecker());
+        const classNode =
+            getDeclaration(program, file.name, 'TestClass', isNamedVariableDeclaration);
```
The current implementation returns `null` if you pass the `TestClass_1` variable declaration as the `ClassDeclaration` parameter to `getBaseClassIdentifier()`. This is the same behaviour as `hasBaseClass()`.

I think it is true to say that you would call `getDeclarationOfIdentifier()` on the `TestClass_1` identifier to get the "real" class declaration before calling `getBaseClassIdentifier()`.

petebacondarwin committed:
test(ivy): ensure that `runInEachFileSystem` cleans up after itself

Previously the last file-system being tested was left as the current
file-system. Now it is reset to an `InvalidFileSystem` to ensure future
tests are not affected.

petebacondarwin committed:
feat(ivy): add `getBaseClassIdentifier()` to `ReflectionHost`

This method will be useful for writing ngcc `Migrations` that
need to be able to find base classes.

petebacondarwin committed:
refactor(ivy): ngtsc - remove unnecessary type on helpers

The `ClassDeclaration` already contains the `{name: ts.Identifier}`
type so there is no need to include it explicitly here.

petebacondarwin committed:
feat(ivy): ngcc - support ngcc "migrations"

This commit implements support for the ngcc migrations
as designed in https://hackmd.io/KhyrFV1VQHmeQsgfJq6AyQ

petebacondarwin committed:
feat(ivy): ngcc - implement `UndecoratedParentMigration`

Implementing the "undecorated parent" migration described in
https://hackmd.io/sfb3Ju2MTmKHSUiX_dLWGg#Design

petebacondarwin committed:
refactor(ivy): ngcc - move decorator analysis types into their own file

alxhub commented on packages/compiler-cli/ngcc/src/migrations/undecorated_parent_migration.ts:
```diff
@@ -0,0 +1,128 @@
+  }
+}
+
+function hasDirectiveDecorator(host: MigrationHost, clazz: ClassDeclaration): boolean {
```
Maybe add a comment here just clarifying this checks if the class has an `@Component` or `@Directive` decorator, since both will be accepted.

alxhub commented on packages/compiler-cli/ngcc/src/migrations/undecorated_parent_migration.ts:
```diff
@@ -0,0 +1,128 @@
+/**
+ * Ensure that a tree of AST nodes have their parents wired up.
+ */
+function setParentPointers(parent: ts.Node, child: ts.Node): void {
```
Move this to a utility method in a common file.

petebacondarwin commented on packages/compiler-cli/ngcc/src/host/esm2015_host.ts:
```diff
@@ -190,9 +190,21 @@ export class Esm2015ReflectionHost extends TypeScriptReflectionHost implements N
+    return super.hasBaseClass(innerClassDeclaration);
+  }
+
+  getBaseClassIdentifier(clazz: ClassDeclaration): ts.Identifier|null {
```
Good call!

I will change this to return an `Expression`, and then the `Migration` instances will need to check it is an `Identifier` before using it (or use the `PartialEvaluator` if the migration can cope with more dynamic scenarios).

petebacondarwin committed:
fixup! feat(ivy): add `getBaseClassIdentifier()` to `ReflectionHost`

petebacondarwin committed:
refactor(ivy): use ReflectionHost to find base classes

When analyzing components, directives, etc we capture its base class.
Previously this assumed that the code is in TS format, which is not
always the case (e.g. ngcc).
Now this code is replaced with a call to
`ReflectionHost.getBaseClassExpression()`, which abstracts the work
of finding the base class.

petebacondarwin committed:
fixup! feat(ivy): ngcc - support ngcc "migrations"

petebacondarwin committed:
fixup! feat(ivy): ngcc - implement `UndecoratedParentMigration`

petebacondarwin commented:
Merge assistance: can a Googler run presubmit please?

mhevery commented:
http://test/OCL:259557101:BASE:259557118:1563901239723:c793e270
